FOCUS #76: Create Datetime Requirements attribute (#123)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: supporting_content/attributes/datetime_requirements.md

# Datetime Requirements

## Example provider mappings

Current column mappings found in available data sets:

| Provider  | Data set                 | Column                                               |
|:----------|:-------------------------|:-----------------------------------------------------|
| AWS       | CUR                      | lineItem/UsageStartDate, lineItem/UsageEndDate, etc. |

> | GCP       | Big Query Billing Export | usage_start_time, usage_end_time, etc.               |

Reviewer: export_time: A processing time associated with an append of Cloud Billing data. This will always increase with each new export.

https://cloud.google.com/billing/docs/how-to/export-data-bigquery-tables/standard-usage

Author: the idea was to list a couple of examples that we map to the dimensions covered by the v0.5 scope. export_time, requires inclusion of fractions of seconds. at some point we might have to include datetime-interval format as well. we will update the specification as the need arises...

Author: As agreed at the squad 2 meeting, for v5.0, only the second level precision format was retained in the specification. note related to other levels of precision (minutes, milliseconds, or microseconds) added to supporting content. We plan to create an additional issue on this topic.

| Microsoft | Cost details             | date, etc.                                           |

### Documentation

> * AWS: ?

Reviewer 2: Do we want to leave this like this? Does it imply that the links are not ready yet OR that they will be added in the future when they are created?

Reviewer 3: I believe these to be the docs that detail the columns that are datetime but they aren't very clear on the format they use or will use.  

* AWS: https://docs.aws.amazon.com/cur/latest/userguide/Lineitem-columns.html#Lineitem-details-U
* GCP: https://cloud.google.com/billing/docs/how-to/export-data-bigquery-tables/standard-usage
* Microsoft: https://learn.microsoft.com/en-us/azure/cost-management-billing/automate/understand-usage-details-fields

Reviewer 2: @ijurica  thoughts?

Author: the idea was to replace the question marks prior to merge to working draft :)

Author: added references to providers documentation

* GCP: ?

> * Microsoft: ?

Reviewer 4: ```suggestion
* Microsoft: https://learn.microsoft.com/azure/cost-management-billing/automate/understand-usage-details-fields
```

Author: done

## Example usage scenarios

Current values observed in billing data for various scenarios:

| Provider  | Data set                                        | Example Value        |
|:----------|:------------------------------------------------|:---------------------|
| AWS       | CUR                                             | 2023-05-13T21:00:00Z |
| GCP       | Big Query Billing Export                        | 2023-05-13T21:00:00Z |
| Microsoft | Cost details via Consumption API (usageDetails) | 2023-05-13T00:00:00Z |
| Microsoft | Cost details via Cost export file               | 05/13/2023           |

## Discussion / Scratch space

> * Date related dimension in Azure Cost exports are not aligned with ISO 8601 and those provided in Consumption Usage Details API are (Note: the API will be retired at some point)

Reviewer 4: ```suggestion
* Date related dimension in Microsoft Cost Management exports are not aligned with ISO 8601 and those provided in Consumption Usage Details API are (Note: the API will be retired at some point)
```

Author: done

* Dates should always have time / timezone to avoid ambiguity. Will result in more space usage but the additional clarity is justified
* Oddities in time / date / timezone

>   * Azure provides daily level data without hourly granularity. Is it wasteful to have to specify the full date/time?

Reviewer 2: This comment could be removed as it was stated that Time would ALWAYS be provided as a rule?
```suggestion
  * Azure currently provides daily-level data without hourly granularity.
```

Reviewer 3: Yes, I agree it should be removed.

Author: suggested changes applied
